src/Addons/CategoryFetcher.php

```diff
@@ -0,0 +1,276 @@
+            ->setStorage($storage)
+        ;
+        $this->factory = new AdvancedCircuitBreakerFactory();
+    }
+
+    /**
+     * @param string $isoCode Two letters iso code to identify the country (ex: en, fr, es, ...)
+     *
+     * @return array
+     */
+    public function getData($isoCode)
+    {
+        $category = $this->getCategoryFromApi($isoCode);
+        $category = $this->addTracking($category, $isoCode);
+        $category['description'] = $this->getDescription($category);
+
+        return $category;
+    }
+
+    /**
+     * @param string $isoCode
+     *
+     * @return array
+     */
+    private function getCategoryFromApi($isoCode)
+    {
+        $circuitBreaker = $this->factory->create($this->apiSettings);
+        $apiJsonResponse = $circuitBreaker->call(
+            self::ADDONS_API_URL . '?iso_lang=' . $isoCode, //Include language in url to correctly cache results
+            function () { return false; },
+            [
+                'body' => [
+                    'method' => 'listing',
+                    'action' => 'categories',
+                    'version' => '1.7',
+                    'iso_lang' => $isoCode,
+                ],
+            ]
+        );
+        $apiResponse = !empty($apiJsonResponse) ? json_decode($apiJsonResponse, true) : false;
+        $category = null;
+        if (false !== $apiResponse && !empty($apiResponse['module']) && empty($apiResponse['errors'])) {
+            $category = $this->searchCategory($apiResponse['module'], $this->categoryId);
+        }
+
+        return null !== $category ? $category : $this->defaultData;
+    }
+
+    /**
+     * @param array $categories
+     * @param int $searchedCategoryId
+     *
+     * @return array|null
+     */
+    private function searchCategory(array $categories, $searchedCategoryId)
+    {
+        foreach ($categories as $category) {
+            if (!empty($category['id_category']) && $searchedCategoryId == $category['id_category']) {
+                return $category;
+            }
+
+            if (!empty($category['categories'])) {
+                $subCategory = $this->searchCategory($category['categories'], $searchedCategoryId);
+                if (null !== $subCategory) {
+                    return $subCategory;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @param array $category
+     *
+     * @return string
+     */
+    private function getDescription(array $category)
+    {
+        $defaultDescription = !empty($this->defaultData['description']) ? $this->defaultData['description'] : '';
+        //Clean link used to fetch description (no need for tracking then)
+        if (empty($category['clean_link'])) {
+            return $defaultDescription;
+        }
+
+        $circuitBreaker = $this->factory->create($this->platformSettings);
+        $categoryResponse = $circuitBreaker->call($category['clean_link'], function () { return false; });
```

Comment:
an interesting feature for the circuit breaker = declare a default callback

Reply:
I'm wondering.. you mean define in the factory settings a default callback? and make the parameter in the call method optional?
Then what would happen if you don't define a default callback nor one in the call method?